[HVG-350] Fix / complete Danish home content quote

styrelius opened the pull request:
### What?

Make sure we both query for and display Danish quote details like street, type of home, number of people that should be covered by the insurance etc.

1. Specify what `DanishHomeContentsDetails` we want to get back from queries `Quote` and `QuoteBundle`, and re-generate the `graphql.tsx` file to get this new stuff in there.
2. Make additions and modifications to `/OfferNew/utils.ts` and `InsuranceSummary.tsx` in order to get the existing logic work with Danish quotes as well as Swedish and Norwegian. 
More specifically the `getHouseholdSize` function in `utils.tsx` is altered to not care which type the quote is, but instead check for the properties `householdSize` and `coInsured` to get the correct size of the household. Also if that function returns `0` then `-` will be rendered instead of a 0 since that seemed to cause confusion when we tested the signup flow.

<br>

### Why?

There was no information in the "Your details" section on the checkout sidebar when fetching a Danish quote.

_Referenced ticket [here](https://hedvig.atlassian.net/browse/HVG-350)_

<br>

### Screenshots

**the correct details under "Your details"**
<img width="1893" alt="2020-11-09-danish-checkout" src="https://user-images.githubusercontent.com/42962286/98577661-91c90b00-22bc-11eb-8b35-56722c12f88d.png">

<br>

### GIFs

**changing number of co-insured and it works!**

![2020-11-09-coInsured-dk](https://user-images.githubusercontent.com/42962286/98578911-35ff8180-22be-11eb-8b92-d234129858cb.gif)

styrelius committed:
Modify 'Quote' query (Quote.graphql) to specify what we want to get back from DanishHomeContentsDetails. Run gen script to get those changes into graphql.tsx.

styrelius committed:
Modify 'QuoteBundle' query (in QuoteBundle.graphql) to specify what data we should get back from DanishHomeContentsDetails. Run gen script to get those changes into graphql.tsx.

styrelius committed:
Fix 'getHouseHoldSize' and 'quoteDetailsHasAddress' functions in /OfferNew/utils.ts so that they work with Danish quotes.
Fix InsuranceSummary.tsx so that it displays all quote details correctly for Danish quotes. Add a condition to ternary deciding the household size value so that it renders '-' if the houseHoldSize is 0 (since it should never be 0 and that means something's wrong).

styrelius committed:
Add TODO's in CheckoutContent.tsx and Sign.tsx - these are things that should probably be fixed before launching in Denmark but they are not critical.

styrelius committed:
Fix spelling in InsuranceSummary.tsx.

styrelius committed:
Remove TODO's since one of them was unnecessary and one I added to our backlog instead.

styrelius committed:
Merge branch 'master' into fix/full-danish-home-content-quote

styrelius committed:
Merge branch 'master' into fix/full-danish-home-content-quote

virrios commented on src/client/pages/OfferNew/Checkout/InsuranceSummary.tsx:
```diff
@@ -226,15 +227,19 @@ const getDetails = (
+        houseHoldSize === 0
+          ? '-'
+          : houseHoldSize === 1
           ? textKeys.CHECKOUT_DETAILS_SINGLE_PERSON()
```
When using the debugger for sweden there is a problem when selecting 0 for house hold size. It seems like it creates a quote (createQuote) , however Query "Quote" generates an error: 
errors: [{message: "Number is required.", locations: [{line: 2, column: 3}], path: ["quote"]}]
locations: [{line: 2, column: 3}]
message: "Number is required."
path: ["quote"]

styrelius replied:
Yes because you shouldn't select 0 - that means the insurance is for no-one! So it's actually really good that it throws an error imo. The thing is in the Swedish onboarding flow we as for the total amount of people the insurance is for (that means, you included - i.e. at least 1 should be submitted here) and in Norway and Denmark we ask for co-insured, that is how many persons _excluding_ you should be covered by the insurance. But maybe it should throw an error already when creating the quote.

styrelius replied:
Maybe we should just throw an error if the `getHouseholdSize` function returns `0`? 🤔

virrios replied:
I mean its a debugger so maybe super prio  (we are the ones using it) but you actually create a quote, so if you try to add a quote with houseHoldSize= 0 nothing happens and then if you change it (without nuking the state because it looks like nothing has happened)  to a valid number and try again it will cause a quiteId already exist error.  

I mean it's more important, as you said to get the entire embark->offer flow and un-block the app team 

But a question  can it ever show '-' if you can't ever select householdSize =0? I am a bit confused here.

styrelius replied:
It's pretty confusing yes, haha! But ultimately I guess it's a backend issue that you're actually able to create a quote for zero persons. 
The thing is we actually got the number 0 for "Number of people" in the checkout when we did the testing session, I don't know if you remember? Due to this being handled strangely. Anyway, I've changed it so it throws an error instead of the number is 0 now, as I  think you saw! 😀 But that will probably never happen now, I mean if you get that error something's really screwed with the backend...

styrelius commented on src/client/pages/OfferNew/Checkout/InsuranceSummary.tsx:
```diff
@@ -197,7 +197,8 @@ const getDetails = (
             quoteDetails.__typename === 'SwedishApartmentQuoteDetails' ||
-            quoteDetails.__typename === 'NorwegianHomeContentsDetails'
+            quoteDetails.__typename === 'NorwegianHomeContentsDetails' ||
+            quoteDetails.__typename === 'DanishHomeContentsDetails'
```
I get your point. I added `quoteDetails.__typename === 'DanishHomeContentsDetails'` here since the `'NorwegianHomeContentsDetails'` was already here and Danish home contents should work as the Norwegian one. I think this is a good thing to take into account when we refactor this but for now I feel like it's more important to get this to work so we can get the entire embark->offer flow and un-block the app team.

styrelius replied:
Or maybe I'm misunderstanding.

virrios replied:
Yeah as you said, its was already like that for NorwayHomeContent. I it was more of a general question but I
 think I might have been a bit unclear what i meant. 
We are using the CHECKOUT_APARTMENT textkey for when we display HomeContentDetails. So my question was if don't have a specifik text key for HomeContentDetails but rather that we use CHECKOUT_APARTMENT for HomeContent as well.

styrelius replied:
Aha, I see! Yes maybe we should change the text key, you have a point. Not gonna do that now though.

palmenhq commented on src/client/pages/OfferNew/Checkout/InsuranceSummary.tsx:
```diff
@@ -226,15 +227,19 @@ const getDetails = (
       label: textKeys.CHECKOUT_DETAILS_HOUSEHOLD_SIZE(),
       value:
-        getHouseholdSize(quoteDetails) === 1
+        householdSize === 0
+          ? '-'
+          : householdSize === 1
           ? textKeys.CHECKOUT_DETAILS_SINGLE_PERSON()
```
i'm not a fan of nested ternaries as they imo get quite hard to read. Could we split this up into variables or named functions maybe?

styrelius replied:
Sure, I can see that you would think that! 😀  I'll see what I can do!

palmenhq commented on src/client/pages/OfferNew/utils.ts:
```diff
@@ -57,13 +58,10 @@ export const getOfferData = (quoteBundle: QuoteBundle): OfferData => {
     return quoteDetails.householdSize
   }
-  if (
-    isNorwegianHomeContents(quoteDetails) ||
-    isNorwegianTravel(quoteDetails)
-  ) {
+  if ('coInsured' in quoteDetails) {
```
🤯  woah

styrelius replied:
😇

palmenhq reviewed (commented):
Awesome!

styrelius committed:
Replace nested/chained ternary in InsuranceSummary.tsx with 'getHouseHoldSizeValue' function.

styrelius committed:
Merge branch 'fix/full-danish-home-content-quote' of github.com:HedvigInsurance/web-onboarding into fix/full-danish-home-content-quote

styrelius commented on src/client/pages/OfferNew/Checkout/InsuranceSummary.tsx:
```diff
@@ -226,16 +227,26 @@ const getDetails = (
+    }
+    return '-' // or should we just throw an error?
+  }
+
```
@palmenhq Is this better?

styrelius commented on src/client/pages/OfferNew/Checkout/InsuranceSummary.tsx:
```diff
@@ -226,16 +227,26 @@ const getDetails = (
     ])
   }
 
+  const getHouseHoldSizeValue = (
```
@palmenhq Is this better?

styrelius replied:
Nice that I made this function really hard to view now by adding comments in the middle of it 🧠

styrelius replied:
Oops I might have deleted your comment by deleting my own comment @palmenhq 😳 What a mess!

palmenhq replied:
Haha its ok 😂

palmenhq commented on src/client/pages/OfferNew/Checkout/InsuranceSummary.tsx:
```diff
@@ -226,16 +227,26 @@ const getDetails = (
+        VALUE: householdSize,
+      })
+    }
+    return '-' // or should we just throw an error?
```
i do think it makes more sense to throw an error, otherwise we'll never know if this actually happens

styrelius replied:
I agree! Will fix that 👍

styrelius committed:
Add some throwing Errors:
1) In 'getHouseHoldSizeValue' function in InsuranceSummary.tsx if householdSize is less than 1.
2) In 'getHouseholdSize' function in /OfferNew/utils.ts if neither property 'householdSize' nor 'coInsured' exists on quoteDetails.

palmenhq reviewed (approved):
amaze! 👑

virrios commented:
Nice!!
